### PR TITLE
Hfp import

### PR DESCRIPTION
**Pull request by Divinare**
Now you can run the api with:
```
./run_api.sh
```
And importer with:
```
run_importer.sh
```
NOTE: to get importer working locally, I had to install cffi, checked instructions from here: https://stackoverflow.com/questions/34370962/no-module-named-cffi-backend
NOTE2: not sure why, but run_importer.sh still sometimes throws error "Exception: ModuleNotFoundError: No module named '_cffi_backend'." and when i re-run it, it just works.

Currently only DOC and DOO are being imported. You can try to import VP's as well by adding VP to this array:
event_types_to_import = ['DOC', 'DOO']

As you can see, I was able to import VP, DOO and DOC events:

[2022-05-19T11:56:16.960Z] Importing blob: 2022-05-18T01-3_DOC.csv.zst
[2022-05-19T11:56:16.960Z] Import invalid row count: 0
[2022-05-19T11:56:16.960Z] Importing blob: 2022-05-18T01-3_VP.csv.zst
[2022-05-19T11:56:16.960Z] Import invalid row count: 0
[2022-05-19T11:56:16.960Z] Importing blob: 2022-05-18T01-4_VP.csv.zst
[2022-05-19T11:56:16.960Z] Import invalid row count: 0
[2022-05-19T11:56:16.961Z] Importing blob: 2022-05-18T02-1_VP.csv.zst
[2022-05-19T11:56:16.961Z] Import invalid row count: 0
[2022-05-19T11:56:16.961Z] Importing blob: 2022-05-18T02-2_DOC.csv.zst
[2022-05-19T11:56:16.961Z] Import invalid row count: 0
[2022-05-19T11:56:16.961Z] Importing blob: 2022-05-18T02-2_VP.csv.zst
[2022-05-19T11:56:16.961Z] Import invalid row count: 0
[2022-05-19T11:56:16.961Z]  imported_successfully True
[2022-05-19T11:56:16.961Z] Import done successfully, running analysis.
[2022-05-19T11:56:16.961Z] 326312 observations inserted.
[2022-05-19T11:56:16.961Z] 515 observations updated with guessed stop_id
[2022-05-19T11:56:16.961Z] 159647 observations updated with dist_to_jore_point_m
[2022-05-19T11:56:16.961Z] 58 rows deleted from "stop_median"
[2022-05-19T11:56:16.961Z] 169 rows inserted into "stop_median"
[2022-05-19T11:56:16.961Z] 116510 observations updated with dist_to_median_point_m
[2022-05-19T11:56:16.961Z] 676 "percentile_radii" created using percentiles 0.5,0.75,0.9,0.95
[2022-05-19T11:56:16.962Z] 156 "stop_median" updated with "result_class", "recommended_min_radius_m" and "manual_acceptance_needed"

NOTE: I got this output when I had this:
```
WHERE (
      hfp.event_type = 'VP' OR 
      hfp.event_type = 'DOO' OR
      hfp.event_type = 'DOC'
    )
```    
in stopcorr.refresh_observation(). But currently we only need DOO and DOC so I removed VP from there. At least it's good to know that VP's can also be imported.

### CODE REVIEW

File: db/sql/150_stop_correspondence_schema.sql [hunk 1]
```diff
@@ -50,6 +52,41 @@ COMMENT ON COLUMN observation.lat IS
...
+      hfp.longitude,
+      hfp.latitude
+    FROM hfp.view_as_original_hfp_event hfp
+    WHERE (
+      hfp.event_type = 'DOO' OR
+      hfp.event_type = 'DOC'
+    ) AND
+    hfp.longitude IS NOT NULL AND
+    hfp.latitude IS NOT NULL
+    ON CONFLICT DO NOTHING
```
Comment (Divinare) [inline db/sql/150_stop_correspondence_schema.sql:81]: Got some primary key conflicts so I added this. I suppose it's because some events happened during the same second.

File: python/importer/main.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        print(f'Error in reading blob chunks: {e}')
+
+    return imported_successfully
+
+def read_imported_data_to_db(cur, downloader):
+    compressed_content = downloader.content_as_bytes()
+    reader = zstandard.ZstdDecompressor().stream_reader(compressed_content)
+    bytes = reader.readall()
+    hfp_dict_reader = csv.DictReader(StringIO(bytes.decode('utf-8')))
+    import_io = StringIO()
```
Comment (Divinare) [inline python/importer/main.py:73]: I had to get rid of reading in chunks as I think the chunk size was different zstd frame so it could not decompress the chunk. Well, one 15 min VP file should be max 300 MB so memory overflow shouldn't be a problem.

File: db/sql/150_stop_correspondence_schema.sql [hunk 3]
```diff
@@ -1,13 +1,15 @@
...
---       are easy to group and identify together on db side.
--- 
+---
+--- Tables for stop correspondence analysis.
+--- TODO: Move all tables under stopcorr schema
+---       This way, objects related to particular business cases
+---       are easy to group and identify together on db side.
+---
+CREATE SCHEMA stopcorr;
+COMMENT ON SCHEMA stopcorr IS
```
Comment (Divinare) [inline db/sql/150_stop_correspondence_schema.sql:8]: Created the schema already but didn't want to move existing tables under it yet. Perhaps better to do it in another PR

File: python/importer/main.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+    invalid_row_count = 0
+    for hfp in hfp_dict_reader:
+        if hfp["vehicleNumber"] != None:
```
Comment (datarttu) [inline python/importer/main.py:77]: I've stumbled upon HFP messages even without `tst` or `operatorId` as well and would find it a good idea to validate existence of all these primary key values here:

```
if not any([hfp[key] is None for key in ["tst", "operatorId", "vehicleNumber"]]:
    # do write csv stuff
else:
    invalid_row_count += 1
```

File: python/importer/main.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+            row = f'{hfp["tst"]},{hfp["eventType"]},{hfp["receivedAt"]},{hfp["ownerOperatorId"]},{hfp["vehicleNumber"]},{hfp["mode"]},{hfp["routeId"]},{hfp["dir"]},{hfp["oday"]},{hfp["start"]},{hfp["oper"]},{hfp["odo"]},{hfp["drst"]},{hfp["locationQualityMethod"]},{hfp["stop"]},{hfp["longitude"]},{hfp["latitude"]}\n'
+            import_io.write(row)
+        else:
+            invalid_row_count += 1
+    # TODO: log invalid row count into db
+    print(f'Import invalid row count: {invalid_row_count}')
```
Comment (datarttu) [inline python/importer/main.py:83]: Here, we could specify the reason for invalid rows to achieve a clearer log entry, e.g. `f'Rows not imported due to missing tst, operatorId, or vehicleNumber value: {invalid_row_count}'`. I'd also consider logging this only conditionally, i.e. `if invalid_row_count > 0`.

File: db/sql/150_stop_correspondence_schema.sql [hunk 6]
```diff
@@ -1,13 +1,15 @@
...
+'For stop correspondence analysis';
 
 --
--- HFP observations
+-- HFP observations (DOC / DOO events)
 --
 CREATE TABLE observation (
```
[the remark this answers is not in the document]
Comment (Divinare) [reply db/sql/150_stop_correspondence_schema.sql:14]: Yeah let's do it in a separate PR

File: db/sql/150_stop_correspondence_schema.sql [hunk 7]
```diff
@@ -1,13 +1,15 @@
--- 
--- Tables for stop correspondence analysis.
--- TODO: Move these tables later to a separate schema,
---       e.g. CREATE SCHEMA stopcorr; CREATE TABLE stopcorr.observation ... etc.
---       This way, objects related to particular business cases
---       are easy to group and identify together on db side.
--- 
+---
+--- Tables for stop correspondence analysis.
+--- TODO: Move all tables under stopcorr schema
```
[the remark this answers is not in the document]
Comment (Divinare) [reply db/sql/150_stop_correspondence_schema.sql:3]: Will remove it in a separate PR

**Review by datarttu — APPROVED** [review]
I'm currently not able to reproduce this branch in my local environment (see #17 ) but given that it works elsewhere, looks good to me! 👍
